target/ppc/fpu_helper.c

```diff
@@ -225,13 +243,33 @@ static void finish_invalid_op_arith(CPUPPCState *env, int op,
     finish_invalid_op_excp(env, op, retaddr);
 }
 
+/* TODO: this is not a 'finish' anymore, find a better name */
+static void finish_invalid_op_arith2(CPUPPCState *env, int op, bool set_fpcc)
+{
+    env->fpscr &= ~(FP_FR | FP_FI);
```

Comment:
From what I understand to set/unset these bits you pass them through the `mask` to `finish_fp_operation`, so I don't understand why this function unset them even if they're not a part of the mask.

Reply:
The ISA says in Invalid Operation Exception section that
> If the operation is an arithmetic, Floating Round to Single-Precision, Floating Round to Integer, or convert to integer operation,
> ...
> FR FI are set to zero

I agree this is a weird side effect hidden in this function. I was thinking this would be necessary to accommodate some cases in the future like when invalid should no set inexact (as was happening here https://github.com/PPC64/qemu/commit/be2e904dd85cb97b090e19b04215431b3f89b5d5#r68851624 but seems to be fixed in a more recent version) but this might not be case. I did some more tests and I think I'll drop this line.

Thanks for your review

Reply:
IIUC, these methods would only be called when the bits they change are in the mask. That means that the knowledge of which bits can be changed is in the caller, but knowledge of which bits will change is in the callee. I think it makes it harder to reason about the code. E.g., if we need to change something in this method, we'll need to check all callers (and the calls to the callers) to see if the change is allowed. 

That's not far from the current situation, but it feels like we can do better. Maybe we shouldn't change `env->fpscr` directly, but return the bits to be changed, and let the caller do a `env->fpscr = (old_fpscr & ~mask) | (new_fpscr & mask)`. The problem then becomes the side effects of the methods, like setting `env->error_code`.

Another option is to follow the FPCC approach and have a `bool set_fifr` argument, which probably supersedes the need for a `mask` argument but may not be generic enough to solve other incorrect-bit-is-setted problems.